microsoft/adaptivecards-templates
# Add basic deploy to azure button

**Tickets**: [TASK-34368](https://microsoftgarage.visualstudio.com/Intern%20GitHub/_workitems/edit/34368), [TASK-34449](https://microsoftgarage.visualstudio.com/Intern%20GitHub/_workitems/edit/34449), [TASK-34414](https://microsoftgarage.visualstudio.com/Intern%20GitHub/_sprints/taskboard/Adaptive%20Cards/Intern%20GitHub/Y20-VW/12?workitem=34414)

### Summary of Changes
**client/**
Removed Config file and react environment variables from the client code. The client now requests the environment variables (app id and redirect uri) from the backend when the page is first loaded. 
**server/**
Added config endpoint for client to request environment variables. 

**adaptivecards-templating-service/**
Added config endpoint for client to request environment variables. Previous string token replaced variables are replaced with environment variables. 
Published to [npm registry](https://www.npmjs.com/package/adaptivecards-templating-service)
**adaptive-templating-service-typescript-node/**
Published to [npm registry](https://www.npmjs.com/package/adaptive-templating-service-typescript-node). 

**Docker**
Uploaded docker image to ACMSRegistry on Azure Container Registry with most recent dev changes and changes in this branch. 

### Testing these Changes
1. Go to [private-templates-service's README](https://github.com/microsoft/adaptivecards-templates/tree/users/grzhan/docker/private-templates-service). Click the blue 'Deploy to Azure' button. 
2. Select the subscription, resource group, and server farm you wish to deploy your website under.
3.  **If you already have an existing AAD app registration you want to use, skip this step**. Open Azure portal and open Azure Active Directory and select App registrations. Create a new registration with your desired site url. 
4. With your existing or newly created AAD app registration, double check that your new website url is listed as a redirect URI.
5. Enter in your existing mongodb instance's connection string, and the credentials to access the acmsregistry.azurecr.io/acmsimage docker image and deploy. 

The deployment process will take around 10 minutes on the basic plan, you can monitor the progress under Monitor/Log Stream or under Container settings. If you find that it is taking over 10 minutes, restart the web app. 

### Running locally
1. If you already have node_modules installed under server or client, run `npm uninstall adaptivecards-templating-service` under server and `npm uninstall adaptive-templating-service-typescript-node` under client. 
2. Follow the instructions [here](https://github.com/microsoft/adaptivecards-templates/tree/dev/private-templates-service) to run the server and client locally. 

### Revised Instructions for Deployment
**Making changes to `adaptivecards-templating-service`**
1. Go under `adaptivecards-templating-service`. Run `npm version {version_number}`, with the version number set to the next incremented number. 
2. Run `npm publish`.
3. Check [adaptivecards-templating-service](https://www.npmjs.com/package/adaptivecards-templating-service) on the npm registry to see that the new version has been released. 

**Making changes to `adaptivecards-templating-client/typescript`**
1. Refer to [generating client SDK instructions](https://github.com/microsoft/adaptivecards-templates/tree/dev/private-templates-service/adaptivecards-templating-client) on how to update the client SDK code. 
1. Go under `adaptivecards-templating-client/typescript`. Run `npm version {version_number}`, with the version number set to the next incremented number. 
2. Run `npm publish`.
3. Check [adaptive-templating-service-typescript-node](https://www.npmjs.com/package/adaptive-templating-service-typescript-node) on the npm registry to see that the new version has been released. 

**Creating a docker image**
_**This step should be done whenever server or client code is updated. This does not need to be done if changes are made under `adaptivecards-templating-service` or `adaptive-templating-service-typescript-node`.**_
1. Go under `private-templates-service`. Remove `adaptivecards-templating-service` and `adaptivecards-templating-client` directories from your workspace. 
2. If in development mode/client sdk and backend packages were relatively linked (using npm link), run `npm unlink adaptivecards-templating-service` under `server` and `npm unlink adaptive-templating-client`. 
3. Under `server` run `npm install adaptivecards-templating-service` to grab the latest version of our backend library from the npm registry. 
4. Under `client` run `npm install adaptive-templating-service-typescript-node` to grab the latest version of our typescript client SDK from the npm registry. 
5. Delete any generated javascript files, including everything under the `dist/` folder under `server`. 
6. [Download Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest).
7. Run `az login` and enter your Microsoft credentials. 
8. Run `az acr build --registry ACMSRegistry --subscription {subscription_id} --image acmsimage .`
9. Check that image has been updated under "Container Registry" on the Azure portal.

### **Other Options Tested for Deploy To Azure Process**
**1. Setting ARM template to clone directly/link from our Github repo**
_Issues_: Running `npm install` on the app service server does not work as server is underpowered and has slow disks, the build never completes and start up will time out after 20 minutes. On the free plan, this way of setting up our site never works, with the basic plan, it rarely works - we mostly likely need a higher app service pricing tier in order to get this way to work without timing out. 

**2. Installing all dependencies and packaging into a zip file**
This is the method we used for our deployment using Github actions. 
_Issues_: Where would we host the zip file? How often would this be generated?

### Next Steps
1. [TASK-34206](https://microsoftgarage.visualstudio.com/Intern%20GitHub/_sprints/taskboard/Adaptive%20Cards/Intern%20GitHub/Y20-VW/12?workitem=34206): Update deployment and running locally documentation
2. Need to test whether the deployment process works on the free tier for app services. 
3. [TASK-34415](https://microsoftgarage.visualstudio.com/Intern%20GitHub/_sprints/taskboard/Adaptive%20Cards/Intern%20GitHub/Y20-VW/12?workitem=34415): Add MongoDB to ARM template. 
4. [TASK-34413](https://microsoftgarage.visualstudio.com/Intern%20GitHub/_sprints/taskboard/Adaptive%20Cards/Intern%20GitHub/Y20-VW/12?workitem=34413): Write documentation to create AAD app registration. Currently I have not found of a way to add app registration to the ARM template, will see if a script is possible.
